### PR TITLE
[NLU-3727] Arabic date and time support

### PR DESCRIPTION
**Pull request by samhickey25**

### CODE REVIEW

File: Python/build.sh [hunk 1]
```diff
@@ -1,16 +1,3 @@
...
-pip install -r ./requirements.txt
-
-echo // Building Resources
-python index.py ../recognizers-choice/resource-definitions.json
-python index.py ../recognizers-number/resource-definitions.json
-python index.py ../recognizers-number-with-unit/resource-definitions.json
-python index.py ../recognizers-date-time/resource-definitions.json
-python index.py ../recognizers-sequence/resource-definitions.json
-
-cd ../..
```
Comment (samhickey25) [inline Python/build.sh:13]: Does anyone have a problem with removing this? We won't need to define regex in the YAML files anymore, can just update the python resources instead

File: Python/libraries/recognizers-date-time/recognizers_date_time/resources/arabic_date_time.py [hunk 2]
```diff
@@ -0,0 +1,796 @@
...
+    MonthSuffixRegex = f'(?<msuf>(?:(in|of|on)\\s+)?({RelativeMonthRegex}|{WrittenMonthRegex}))'
+    DateUnitRegex = f'((?<unit>(((ال)?(يوم(ا)?|أسبوع(ا)?|شهر(ا)?|سنة|عام(ا)?|قرن|حقبة))|نهاية الأسبوع))|(?<plural>((ال)?(يومان|أسبوعان|شهران|سنتان|عامان|قرنان|حقبتان|يومين|أسبوعين|شهرين|سنتين|عامين|قرنين|حقبتين|يومان|أسبوعان|شهران|سنتان|عامان|قرنان|حقبتان|أيام|أسابيع|أشهر|سنوات|أعوام|حقبات|قرون|سنين|شهور)))|((?<=\\s+\\d{{1,4}})[ymwd]))\\b'
+    DateTokenPrefix = 'في '
+    TimeTokenPrefix = 'عند '
+    TokenBeforeDate = 'في '
+    TokenBeforeTime = 'عند '
+    HalfTokenRegex = f'^(النصف|نصف|والنصف|ونصف)'
+    QuarterTokenRegex = f'^(ربع|الربع|وربع|والربع|إلا ربع|إلا الربع)'
+    ThreeQuarterTokenRegex = f'^(وثلاثة أرباع|ثلاثة أرباع|إلا الربع)'
+    ToTokenRegex = f'\\b(إلا)'
```
Comment (samhickey25) [inline Python/libraries/recognizers-date-time/recognizers_date_time/resources/arabic_date_time.py:59]: Modified this from` \\b(إلا)$` as the $ doesn't really agree with Arabic in my tests

File: Python/libraries/recognizers-date-time/recognizers_date_time/resources/arabic_date_time.py [hunk 3]
```diff
@@ -37,7 +37,7 @@ class ArabicDateTime:
     SpecialDescRegex = f'((?<ipm>)p\\b)'
     AmDescRegex = f'(في\\s)?(صباح(ا)?|صباحًا|الصباح|{BaseDateTime.BaseAmDescRegex})'
     PmDescRegex = f'(في\\s)?((ال)?مساء|مساءً|ليلًا|ليلا|(ال)?ليل(ة)?|بعد الظهر|الظهر|ظهرا|{BaseDateTime.BasePmDescRegex})'
-    AmPmDescRegex = f'(في\\s)?(صباح(ا)?|صباحًا|الصباح|(ال)?مساء|مساءً|{BaseDateTime.BaseAmPmDescRegex})'
+    AmPmDescRegex = f'({BaseDateTime.BaseAmPmDescRegex})'
```
Comment (samhickey25) [inline Python/libraries/recognizers-date-time/recognizers_date_time/resources/arabic_date_time.py:40]: Changed from `(في\\s)?(صباح(ا)?|صباحًا|الصباح|(ال)?مساء|مساءً|{BaseDateTime.BaseAmPmDescRegex})` to allow for am/pm resolution

File: Python/build.sh (hunk 1 above)
Comment (colm-dillon) [reply Python/build.sh:13]: Where is it used? I'd prob hold off on deleting anything for awhile. Maybe a refactoring effort on our fork is required at some point to clean up anything like this we don't use.

**Review by colm-dillon — APPROVED** [review]

File: Python/build.sh (hunk 1 above)
Comment (samhickey25) [reply Python/build.sh:13]: This script is called when we build recognisers. It generates the regexes from the yaml and then runs the test.
I've removed the call to generate the regexes to save us having to define them twice

Comment (andrew-gradinari) [reply Python/build.sh:13]: If we merge some changes from the MS repo, we'll have to remember to manually generate the resource files in order to incorporate those changes, right?

**Review by andrew-gradinari — APPROVED** [review]
Looks good 👍

File: Python/tests/test_runner_datetime.py [hunk 4]
```diff
@@ -69,6 +71,10 @@ def test_datetime_parser(
         simple_parser_assert(spec_info, actual, expected, 'text', 'Text', True)
         simple_parser_assert(spec_info, actual, expected, 'type', 'Type')
 
+        if ignore_resolution:
```
Comment (kevinwalshgen) [inline Python/tests/test_runner_datetime.py]: I think we should just use the `python` key in `NotSupported` / `NotSupportedByDesign`, in that way we can run tests which are not included in `.Net` implementation but are implemented in python. Currently some of of the tests you have removed python from have the ignore resolution and therefore wouldn't test the results.

File: Python/build.sh (hunk 1 above)
Comment (samhickey25) [reply Python/build.sh:13]: Yeah that's right. We haven't been merging anything though

File: Python/libraries/recognizers-date-time/recognizers_date_time/date_time/arabic/time_parser_config.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+        self._utility_configuration = config.utility_configuration
+        self._numbers: Dict[str, int] = config.numbers
+        self._time_zone_parser = config.time_zone_parser
+
+    def adjust_by_prefix(self, prefix: str, adjust: AdjustParams):
+        delta_min = 0
+        trimmed_prefix = prefix.strip().lower()
+
+        if regex.search(self.half_token_regex, prefix):
```
Comment (kevinwalshgen) [inline Python/libraries/recognizers-date-time/recognizers_date_time/date_time/arabic/time_parser_config.py]: Should this be `trimmed_prefix` here?

**Review by kevinwalshgen — COMMENTED** [review]
More to do, but lunch is calling

File: Python/libraries/recognizers-date-time/recognizers_date_time/date_time/arabic/time_parser_config.py (hunk 5 above)
Comment (samhickey25) [reply Python/libraries/recognizers-date-time/recognizers_date_time/date_time/arabic/time_parser_config.py]: yup good spot

File: Python/tests/test_runner_datetime.py (hunk 4 above)
Comment (samhickey25) [reply Python/tests/test_runner_datetime.py]: actually yeah I agree with you. I don't like this ignore resolution field

**Review by kevinwalshgen — APPROVED** [review]

**Pull request merged**